### PR TITLE
Add class hierarchy for monetary amounts

### PR DESCRIPTION
**Pull request by danhper**
First shot at adding statically typed monetary amounts.  
Still need to add arithmetic operations.

### CODE REVIEW

**Comment by danhper** [comment]
Failing tests are most likely not related to this PR

File: lib/monetary.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+export const BTC = new Bitcoin();
+export const ETH = new Ethereum();
+
+export class ERC20 implements ICurrency {
+    private _name: string;
+    private _decimals: number;
+
+    constructor(name: string, decimals: number) {
```
Comment (nud3l) [inline lib/monetary.ts]: Possibly we could make the `decimals` optional and default to 18.

File: lib/monetary.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
+import Big from 'big.js';
```
Comment (nud3l) [inline lib/monetary.ts:1]: I think we should use BigNumbers from ethers as we are depending on that in other places as well (see https://docs.ethers.io/v5/api/utils/bignumber/). What do you think?

**Comment by nud3l** [comment]
> Failing tests are most likely not related to this PR

Yes, it seems like the 1000 ms in the test setup are too short with the additional tests. Probably increasing that should fix the issue.

**Comment by gregdhill** [comment]
Can we do without jest, the existing tests all use mocha / chai so I think it makes sense to just use one testing framework? What do you think?

File: lib/monetary.ts (hunk 2 above)
Comment (gregdhill) [reply lib/monetary.ts:1]: I think we will need this actually to handle decimals for the front-end, as far as I know `BigNumber` doesn't work for this.

File: lib/monetary.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+
+type BigOrNum = Big | number;
```
Comment (gregdhill) [inline lib/monetary.ts]: Could we support `BigNumberish` from `ethers`?

File: lib/test/monetary.test.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
+import Big from 'big.js';
```
Comment (gregdhill) [inline lib/test/monetary.test.ts]: This file should be moved to `test/lib/monetary.test.ts`.

File: lib/monetary.ts (hunk 3 above)
Comment (danhper) [reply lib/monetary.ts]: Changed to `BigSource` from `big.js`  
I guess changing to BigNumberish will depend on whether we switch to ethers or not.

**Comment by danhper** [comment]
> Can we do without jest, the existing tests all use mocha / chai so I think it makes sense to just use one testing framework? What do you think?

Changed the test runner to mocha.

File: lib/test/monetary.test.ts (hunk 4 above)
Comment (danhper) [reply lib/test/monetary.test.ts]: I would rather keep the library as self-contained as possible and have the library tests live with the library. What do you think?

Comment (danhper) [reply lib/test/monetary.test.ts]: Moved as discussed during the call.

File: lib/monetary.ts (hunk 2 above)
Comment (danhper) [reply lib/monetary.ts:1]: Sticking with big.js for now to support decimals.

File: lib/monetary.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
+import Big from 'big.js';
+import { BigSource } from 'big.js';
+
+
+export interface ICurrency {
```
Comment (gregdhill) [inline lib/monetary.ts:5]: Typescript doesn't require separate interface names, if a class with the same name inherits this type then the compiler should be able to differentiate.

**Review by gregdhill — APPROVED** [review]

**Pull request merged**